src/util/irep2.cpp

```diff
@@ -763,23 +763,20 @@ type_poolt &type_poolt::operator=(type_poolt const &ref)
   return *this;
 }
 
-// XXX investigate performance implications of this cache
+// TODO: investigate performance implications of this cache
 static const type2tc &
-get_type_from_pool(const typet &val, std::map<typet, type2tc> &)
+get_type_from_pool(const typet &val, std::map<typet, type2tc> &map)
 {
-#if 0
-  std::map<const typet, type2tc>::const_iterator it = map.find(val);
-  if (it != map.end())
-    return it->second;
-#endif
-
+  /* BUG: This is creating problems
+  auto it = map.find(val);
+  if(it != map.end())
+    return map[val];
+*/
   type2tc new_type;
   real_migrate_type(val, new_type);
-#if 0
   map[val] = new_type;
+
   return map[val];
-#endif
-  return *(new type2tc(new_type));
 }
 
```

Comment:
You are adding to the map but never reading from it?

What's the point of the change here?

Reply:
The hashing algorithm is not working properly. This means that sometimes the caching is causing bugs.

> What's the point of the change here?

The shared_ptr reference is being held by the map. Which means that it will only be deleted when the index is replaced or when the object dies which removes the memory leak.

Reply:
> > Is this rebased on top of master? The results from master that Lucas posted don't show 132 false negative verdicts.
> > src/util/irep2.cpp
> 
> At the time I started running the experiments the fix wasn't there.

Please update the branch so we can have accurate results.

Reply:
> The hashing algorithm is not working properly. This means that sometimes the caching is causing bugs.
> 
> > What's the point of the change here?
> 
> The shared_ptr reference is being held by the map. Which means that it will only be deleted when the index is replaced or when the object dies which removes the memory leak.

That doesn't make sense: by keeping the type stored on the map, its counter will never reach zero and will never be freed. Can you provide any evidence that the memory is being freed? That this change actually reduces memory usage.

Furthermore, if you identified an issue with the hashing algorithm, that should be fixed. What's the issue you identified?

Reply:
> That doesn't make sense: by keeping the type stored on the map, its counter will never reach zero and will never be freed. 

It will after the map object is deleted. It will call the `delete` over its members. There are two references one of the initial object and another hold by the map, both are removed. The example below with a map shouldn't contain any leaks and it is the same approach:

```c++
std::map<int,std::shared_ptr<int>> m;

void foo() {
    for(int i = 0; i < 5; i++) {
        std::shared_ptr<int> value(new int(i*2));
        m[i] = value;
    }
}

int main() {
    foo();
    for(int i = 0; i < 5; i++) assert(*m[i] == i*2);
    return 0;
}
```

This can be tested with both `valgrind` and `asan`.

> Can you provide any evidence that the memory is being freed? That this change actually reduces memory usage.

I can only show that this removes a memory leak during the build process (on the CI). Not that it reduces the memory usage of esbmc

Remember that it all began with the ASAN build (which enables leak checks on Unix by default). The build is triggering the error because of the `return new...` .

ASAN not complaining anymore should be an evidence that at least it is removing a leak.